overlord/notices/notices.go

```diff
@@ -0,0 +1,333 @@
+		Data: map[string]string{
+			"old-status": old.String(),
+			"new-status": new.String(),
+		},
+	})
+}
+
+func noticeUniqueKey(noticeType NoticeType, key string) string {
+	return fmt.Sprintf("%s-%s", noticeType, key)
+}
+
+func updateNotice(notice *Notice, opts *NoticeOptions) bool {
+	now := timeNow().UTC()
+	var notify bool
+	if notice.repeatAfter == 0 || now.After(notice.lastRepeated.Add(notice.repeatAfter)) {
+		// Update last repeated time if repeat-after time has elapsed (or is zero)
+		notice.lastRepeated = now
+		notice.lastData = opts.Data
+		notify = true
+	}
+	return notify
+}
+
+func newNotice(noticeType NoticeType, opts *NoticeOptions) *Notice {
+	now := timeNow().UTC()
+	return &Notice{
+		noticeType:    noticeType,
+		key:           opts.Key,
+		lastData:      opts.Data,
+		firstOccurred: now,
+		lastOccurred:  now,
+		lastRepeated:  now,
+		repeatAfter:   opts.repeatAfter,
+		expireAfter:   NoticeExpireAfter,
+	}
+}
+
+func (m *NoticeManager) addNotice(noticeType NoticeType, opts *NoticeOptions) error {
+	if opts == nil {
+		return fmt.Errorf("internal error: notice options must be provided")
+	}
+
+	key := noticeUniqueKey(noticeType, opts.Key)
+	notify := true
+	if notice := m.notices[key]; notice != nil {
+		notify = updateNotice(notice, opts)
+	} else {
+		m.notices[key] = newNotice(noticeType, opts)
+	}
+	if notify {
+		m.notifyWaiters(m.notices[key])
+	}
+	return nil
+}
+
+func (m *NoticeManager) flattenNotices(filter NoticeFilter) []*Notice {
+	now := timeNow().UTC()
+	var notices []*Notice
+	for _, n := range m.notices {
+		if n.expired(now) || !filter.matches(n) {
+			continue
+		}
+		notices = append(notices, n)
+	}
+	return notices
+}
+
+func (m *NoticeManager) noticesMatchingFilter(filter NoticeFilter) []*Notice {
+	notices := m.flattenNotices(filter)
+	sort.Slice(notices, func(i, j int) bool {
+		return notices[i].lastRepeated.Before(notices[j].lastRepeated)
+	})
+	return notices
+}
+
+func (m *NoticeManager) notifyWaiters(notice *Notice) {
+	for _, w := range m.waiters {
+		if w.filter.matches(notice) {
+			w.notify(notice)
+		}
+	}
+}
+
+// addNoticeWaiter adds a notice-waiter with the given filters. Processing
+// notices for this waiter stops when the done channel is closed.
+func (s *NoticeManager) addNoticeWaiter(filter NoticeFilter) (ch chan *Notice, waiterId int) {
+	s.lastWaiterID++
+	waiter := &noticeWaiter{
+		filter: filter,
+		// Create a buffered channel, with room for 16 notices. It's far more
+		// than what is probably necessary, but we don't want the waiter to miss
+		// anything, even if there should be a short notice burst.
+		ch: make(chan *Notice, 16),
+	}
+	s.waiters[s.lastWaiterID] = waiter
+	return waiter.ch, s.lastWaiterID
+}
+
+// removeNoticeWaiter removes the notice-waiter with the given ID.
+func (s *NoticeManager) removeNoticeWaiter(waiterId int) {
+	delete(s.waiters, waiterId)
+}
+
+func (m *NoticeManager) wait(ctx context.Context, filter NoticeFilter) ([]*Notice, error) {
```

Comment:
I kept this method pretty identical to the one from pebble as this is the polling approach, as I remember we agreed to keep it like. The 'streaming' approach was for the other endpoint we wanted right?